File: src/main/java/com/github/dkoval/leetcode/challenge/CoinChange2.java
package com.github.dkoval.leetcode.challenge;

/**
 * <a href="https://leetcode.com/problems/coin-change-2/">Coin Change 2</a>
 * <p>
 * You are given an integer array coins representing coins of different denominations and an integer amount representing a total amount of money.
 * <p>
 * Return the number of combinations that make up that amount. If that amount of money cannot be made up by any combination of the coins, return 0.
 * <p>
 * You may assume that you have an infinite number of each kind of coin.
 * <p>
 * The answer is guaranteed to fit into a signed 32-bit integer.
 */
public interface CoinChange2 {

    int change(int amount, int[] coins);

    // O(N * A) time | O(N * A) space, where N = len(coins), A - amount
    // Resource: https://www.youtube.com/watch?v=zAnD3AVufsI
    class CoinChange2TopDown implements CoinChange2 {

        @Override
        public int change(int amount, int[] coins) {
            int n = coins.length;
            return change(amount, coins, 0, new Integer[n][amount + 1]);
        }

        private int change(int amount, int[] coins, int idx, Integer[][] memo) {
            // base cases
            if (amount == 0) {
                return 1;
            }

            if (amount < 0 || idx >= coins.length) {
                return 0;
            }

            // already solved?
            if (memo[idx][amount] != null) {
                return memo[idx][amount];
            }

            // Option #1: include coins[idx]. Note that idx is not incremented since we have infinite number of coins[idx] at our disposal.
            int incl = change(amount - coins[idx], coins, idx, memo);
            // Option #2: exclude coins[idx].
            int excl = change(amount, coins, idx + 1, memo);

            // case solution
            memo[idx][amount] = incl + excl;
            return memo[idx][amount];
        }
    }
}

File: src/main/kotlin/com/github/dkoval/leetcode/challenge/CoinChange2.kt
package com.github.dkoval.leetcode.challenge

/**
 * [Coin Change 2](https://leetcode.com/explore/challenge/card/june-leetcoding-challenge/539/week-1-june-1st-june-7th/3353/)
 *
 * You are given coins of different denominations and a total amount of money. Write a function to compute the number of
 * combinations that make up that amount. You may assume that you have infinite number of each kind of coin.
 */
object CoinChange2 {

    // Resources:
    // https://www.educative.io/edpresso/coin-change-problem-2-finding-the-number-of-ways-to-make-a-sum
    // https://www.youtube.com/watch?v=Nvrhx4lbfLI
    fun change(amount: Int, coins: IntArray): Int {
        val dp = Array(coins.size + 1) { IntArray(amount + 1) }
        // initializing 1st column of the `dp` 2D-array to 1
        // because zero-amount can be made in one possible way: {0}
        for (i in 0..coins.size) {
            dp[i][0] = 1
        }
        // recursive solution
        for (i in 1..coins.size) {
            for (requiredAmount in 0..amount) {
                dp[i][requiredAmount] = if (coins[i - 1] > requiredAmount) {
                    // can't pick the highest coin
                    dp[i - 1][requiredAmount]
                } else {
                    // pick the highest coin
                    val remainingAmount = requiredAmount - coins[i - 1]
                    dp[i - 1][requiredAmount] + dp[i][remainingAmount]
                }
            }
        }
        return dp[coins.size][amount]
    }
}